Import OCID data from cell_towers.csv

theirix opened the pull request:
---
#### Agreements
- [x] **I have reviewed and accepted the [guidelines for contributing](https://github.com/CellularPrivacy/Android-IMSI-Catcher-Detector/blob/development/.github/CONTRIBUTING.md) to this project.**
- [x] **I have reviewed and closely followed the [Style Guide](https://github.com/CellularPrivacy/Android-IMSI-Catcher-Detector/wiki/Style-Guide) for this Android app.**

---
#### Overview

Allow to import locally downloaded OCID data from cell_towers.csv.gz

---
#### Classification
- [ ] Bugfix (non-breaking change which fixes an existing issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor (restructuring of existing code without changing its external functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

_Note_
Also slightly refactored RealmHelper without change of functionality.

---
#### References

PR possibly fixes an issue #435 .

Related to discussion in issue #854 about cell_towers.csv .

---
#### Screenshots

![screenshot](https://cloud.githubusercontent.com/assets/481351/17673166/9f3f88b6-6328-11e6-9907-2407514eeed4.png)
1. Added a new menu item "Import CellTowers Data". Choosing this menu launches a standard file chooser intent.
2. A few new toasted notifications added (not shown).
#### Patch description

_Rationale_
OCID allows to download a full daily dump of cell
data in cell_towers.csv.gz or daily incremental data.
Due to bad stability of OCID API it is useful to be able
to locally import cell_towers.csv.

_Proposed changes_
1. Added a drawer menu "Import CellTowers Data" which launches an intent
   to choose a CSV file or gzipped CSV file with CellTowers data.
2. Async loader in ImportTask that reads a full dump and filters
   it locally with MCC, MNC, cell location just like RequestTask does
   by the API call.
3. ImportTask converts data from CellTowers format (documented) to
   the OCID API CSV format and inserts it to the database with help of
   the RealmHelper.
4. RealmHelper is slightly refactored to allow directly insert data
   without an intermediate file.
5. Added a wrapper around GZIPInputStream because Android implementation
   of GZIPInputStream cannot read files larget than 2 GiB.

_Testing_
Tested on Motorola XT1562, Android 6.0.1.
Import from cell_towers.csv.gz fetches approx 1000 records (Moscow) from
29 millions total and took about twenty minutes on debug version.
Unarchived version is much faster.

theirix committed:
Import OCID data from cell_towers.csv

Rationale.
OCID allows to download a full daily dump of cell
data in cell_towers.csv.gz or daily incremental data.
Due to bad stability of OCID API it is useful to be able
to locally import cell_towers.csv.

Proposed changes:
1. Added a drawer menu "Import CellTowers Data" which launches an intent
   to choose a CSV file or gzipped CSV file with CellTowers data.
2. Async loader in ImportTask that reads a full dump and filters
   it locally with MCC, MNC, cell location just like RequestTask does
   by the API call.
3. ImportTask converts data from CellTowers format (documented) to
   the OCID API CSV format and inserts it to the database with help of
   the RealmHelper.
4. RealmHelper is slightly refactored to allow directly insert data
   without an intermediate file.
5. Added a wrapper around GZIPInputStream because Android implementation
   of GZIPInputStream cannot read files larget than 2 GiB.

Testing.
Tested on Motorola XT1562, Android 6.0.1.
Import from cell_towers.csv.gz fetches approx 1000 records (Moscow) from
29 millions total and took about twenty minutes on debug version.
Unarchived version is much faster.

theirix committed:
Fixed code with checkstyle

theirix commented:
Sorry, forgot to run checkstyle.

One error remains in `MainActivity#getUriDataColumn` about `Cursor` type usage. This type is used in a code for extracting file path from the content URI and I am not sure it can be avoided.

ProGamerGov commented:
Could feature be used with local Wigle databases as well? https://github.com/CellularPrivacy/Android-IMSI-Catcher-Detector/issues/751

The collected data is pretty much the same as far as can tell and as I describe in the linked issue, the app can output the database in csv format.

theirix commented:
> Could feature be used with local Wigle databases as well? #751
> The collected data is pretty much the same as far as can tell and as I describe in the linked issue, the app can output the database in csv format.

Surely, why not. This PR provides an `ImportTask` class that can be also used for file import from multiple formats.

I think I can implement it after the initial review by maintainers.

larsgrefer commented:
@theirix please rebase on the current `development` branch

theirix commented:
@larsgrefer done, now it builds without errors

larsgrefer commented on AIMSICD/src/main/java/com/secupwn/aimsicd/ui/activities/MainActivity.java:
```diff
@@ -255,6 +264,10 @@ void selectDrawerItem(int position) {
             downloadBtsDataIfApiKeyAvailable();
+        } else if (selectedItem.getId() == DrawerMenu.ID.APPLICATION.IMPORT_CELL_TOWERS_DATA) {
+            Intent pickFileIntent = new Intent(Intent.ACTION_GET_CONTENT);
+            pickFileIntent.setType("*/*");
```
I think the category `CATEGORY_OPENABLE` should be added here, so you can only select real files

theirix replied:
Nice idea, added this.

larsgrefer commented:
Why are you trying to get the real file path from the content uri?

Have you considered to use `ContentResolver.openInputStream(uri)`? https://developer.android.com/reference/android/content/ContentResolver.html#openInputStream(android.net.Uri)

theirix committed:
Added CATEGORY_OPENABLE for import file intent

theirix committed:
Open file with ContentResolver::openFileDescriptor

Simplified opening an imported file by using ParcelFileDescriptor
from ContentResolver::openFileDescriptor. Check for gzip is performed
against URI string.

theirix committed:
Switched to ContentResolver::openInputStream

theirix commented:
> Why are you trying to get the real file path from the content uri?
> 
> Have you considered to use ContentResolver.openInputStream(uri)? https://developer.android.com/reference/android/content/ContentResolver.html#openInputStream(android.net.Uri)

You are right, ContentResolver::openInputStream`is more than enough. I implemented a MIME check for gzip/csv by analyzing`ContentResolver::getType` and an URI string. Initially it needed a real path extraction. Latest commits simplified the PR a lot, I can squash them if you want.

theirix committed:
Fixed checkstyle warning

larsgrefer commented on AIMSICD/src/main/java/com/secupwn/aimsicd/ui/activities/MainActivity.java:
```diff
@@ -281,6 +290,27 @@ void selectDrawerItem(int position) {
+                    String type = getContentResolver().getType(data.getData());
+                    boolean isGzip = type != null && type.equals("application/octet-stream") &&
+                            data.getDataString().endsWith(".gz");
+                    InputStream importFile = getContentResolver().openInputStream(data.getData());
```
Im not sure if this is the right place to open the `InputStream` and to check the type.

I'd prefer to pass `data.getData()` around instead of the open `InputStream`

larsgrefer commented on AIMSICD/src/main/java/com/secupwn/aimsicd/utils/ImportTask.java:
```diff
@@ -0,0 +1,284 @@
+            // Prepare filtering values
+            final String mccFilter = String.valueOf(mobileCountryCode);
+            final String mncFilter = String.valueOf(mobileNetworkCode);
+            double earthRadius = 6371.01; // [Km]
```
This one shoud be a constant

theirix committed:
Pass plain Uri to ImportTask

theirix committed:
Log ImportTask inputs

theirix committed:
Made EARTH_RADIUS a constant

theirix commented:
@larsgrefer thank you for the advice, I replaced InputStream to the Uri object, it seems clearer this way. Also refactored a constant.

theirix commented:
Thanks!

sheltonr2 commented:
Is this version with the import option something that can be currently downloaded?

theirix commented:
@sheltonr2 , the latest release 0.1.43 on May 2, 2016 does not include this feature. It should be built from sources.

jayrmartinignaciojr commented:
Nice
